xapian-applications/omega/diritor.h

```diff
@@ -145,8 +145,13 @@ class DirectoryIterator {
 		break;
 	    case DT_REG:
 		return REGULAR_FILE;
-	    case DT_DIR:
-		return DIRECTORY;
+	    case DT_DIR: {
+			char * dot_ptr = strrchr(entry->d_name, '.');
+			std::vector<std::string> dasfile {".key", ".numbers", ".pages", "\0"};
+			if(std::find(dasfile.begin(), dasfile.end(), dot_ptr) != dasfile.end())
+				return REGULAR_FILE;
+			else return DIRECTORY;
+		}
```

Comment:
The indentation here doesn't match the surrounding code.

`if(...` doesn't match the rest of the code (Xapian coding style is to write `if (` with a space between).

It's cleaner to have the statement after the `else` on its own line when the statement after `if` is.

If a directory name doesn't contain a `.` (which is probably the most common case) then `dot_ptr` will be `NULL` and you'll end up implicitly trying to convert that to `std::string`, which I think is undefined behaviour.

Why the `"\0"` entry in `dasfile`?  You're including the `.` in the string searched, so I can't see how that can ever match.  A filename can't contain a zero byte so even if the `.` weren't included I can't see how it could validly match.

This also hard-codes the extensions which get this special handling, which seems undesirable.  Really this should be configurable in a similar way to how indexing of files is configured.

The `vector` here will get created each time a directory is encountered, which is a bit inefficient.  It also requires a linear search, which is OK when it's hard-coded and has 3 (or 4) entries, but if this were actually configurable it would scale unnecessarily poorly as the number of such extensions grew large.

This code won't kick in for filing systems which set `entry->d_type` to `DT_UNKNOWN`.  There we need to call `stat()` to find out if this is a directory or a file or whatever.  The most commonly used Linux filing systems set `d_type` usefully (and that means we can sometimes avoid the call to `stat()` which is a small optimisation) but we can't rely on that.

I'm a bit unsure about the approach of faking these directories as `REGULAR_FILE` at this level.  The `DirectoryIterator` class is conceptually a mostly generic class for iterating a directory, so it seems a bad fit for it to have to care about what we're doing with the contents of the directory - that your patch hard codes a list of extensions into it is an indication of that.  That knowledge really belongs at a higher level where similar knowledge about files already is.

So to me it seems more natural to patch this handling in at the point where `DirectoryIterator::DIRECTORY` is handled.  As a bonus, there's only one place that consumes `DirectoryIterator::DIRECTORY` but more than one place where that's returned (so a complete version of your approach would need to add the same handling to multiple places).